File: speedtest.go
package main

/*
speed test
qty 2 prefilled array size [12k, 12k] multiplying into the first,
20 iterations
consistent result:
reg loop: 8930 9050 9230
goroutines: 2070 2280 2300 2320 2330
*/

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

const (
	numLoops   int = 20
	sizeXArray int = 12000
	sizeYArray int = 12000
)

func process(aArr *[][]float64, bArr *[][]float64, wg *(sync.WaitGroup)) {
	// inside goroutine, loop thru length of this
	for y := 0; y < len(*aArr); y++ {
		//fmt.Println((*arr)[y])
		for x := 0; x < len((*aArr)[y]); x++ {
			(*aArr)[y][x] *= (*bArr)[y][x]
			//fmt.Println((*aArr)[y][x])
		}
	}
	//fmt.Println((*aArr)[0][0]);
	wg.Done()
}

func postAdd(aArr *[][]float64) {
	// inside goroutine, loop thru length of this
	for y := 0; y < len(*aArr); y++ {
		//fmt.Println((*arr)[y])
		for x := 0; x < len((*aArr)[y]); x++ {
			(*aArr)[y][x] += 1.2
			//fmt.Println((*arr)[y][x])
		}
	}
}

func prefill(arr *[][]float64, wg *(sync.WaitGroup)){
	for y := 0; y < len(*arr); y++ {
		//fmt.Println((*arr)[y])
		for x := 0; x < len((*arr)[y]); x++ {
			(*arr)[y][x] = (rand.Float64() * 2) + .0001
			//fmt.Println((*arr)[y][x])
		}
	}
	wg.Done()
}

func main() {
	// start wait group as a pointer
	wg := new(sync.WaitGroup)
	
	// make 2 2d arrays "slices"
	//processArray := make([]float64, sizeXArray)
	aArr := make([][]float64, (sizeYArray))
	bArr := make([][]float64, (sizeYArray))
	for y := 0; y < sizeYArray; y++ {
		aArr[y] = make([]float64, sizeXArray)
		bArr[y] = make([]float64, sizeXArray)
	}
	
	// prefill
	wg.Add(2)
	go prefill(&aArr, wg)
	go prefill(&bArr, wg)
	wg.Wait()
	fmt.Println("done with prefill");
	/*
	fmt.Println(aArr[0][0]);
	fmt.Println(bArr[0][0]);
	*/


	// start timer, process
	timerNow := time.Now()
	for i := 0; i < numLoops; i++ {
		wg.Add(1)
		process(&aArr, &bArr, wg)
	}
	
	// wait on waitgroup then stop timer
	wg.Wait()
	timerNowEnd := time.Now()
	timerStart := (timerNow.Second() * 1e9) + timerNow.Nanosecond()
	timerEnd := (timerNowEnd.Second() * 1e9) + timerNowEnd.Nanosecond()
	timerFinal := (timerEnd - timerStart) / 1000000
	//fmt.Println(aArr[11000][11000])
	fmt.Println(timerFinal, "ms")
	
	// added a post timer function to make sure it isn't cheating
	postAdd(&aArr)
	postAdd(&bArr)
	fmt.Println(aArr[11000][11000])
}

File: speedtest.cpp
#include <iostream>
#include <chrono>
#include <random>

/*
speed test
qty 2 prefilled array size [12k, 12k] multiplying into the first,
20 iterations
consistent result: 5480 5760 5950 6000 6160
*/

int const SIZE_X_ARRAY = 12000;
int const SIZE_Y_ARRAY = 12000;
int const NUM_ITERATIONS = 10;

// seed random from c++11 random
std::random_device rd;
std::mt19937 mt(rd());
std::uniform_real_distribution<double> dist(.0001, 2.000);


// run this process for num iterations
void process(double* aPtrs[], double* bPtrs[]){
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            aPtrs[y][x] *= bPtrs[y][x];
        }
    }
}

void prefill(double* arrayPtrs[]){
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            //printf("%p\n", &(arrayPtrs[y][x])); // retrieve the address [y][x]
            arrayPtrs[y][x] = dist(mt);
        }
    }
}

int main(){
    // 2 new pointer arrays of pointers
    double* aPtrs[SIZE_Y_ARRAY];
    double* bPtrs[SIZE_Y_ARRAY];
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        aPtrs[y] = new double[SIZE_X_ARRAY];
        bPtrs[y] = new double[SIZE_X_ARRAY];
    }
    
    // prefill
    prefill(aPtrs);
    prefill(bPtrs);

    // timer init
    using std::chrono::high_resolution_clock;
    using std::chrono::duration_cast;
    using std::chrono::duration;
    using std::chrono::milliseconds;

    // start timer and loop
    auto t1 = high_resolution_clock::now();
    for(int i = 0; i < NUM_ITERATIONS; i++){
        process(aPtrs, bPtrs);
    }
    
    // get end timer and print out
    auto t2 = high_resolution_clock::now();
    auto ms_int = duration_cast<milliseconds>(t2 - t1);
    std::cout << ms_int.count() << "ms\n";

    // debug test if it actually did the work
    /*
    for(int y = 0; y < SIZE_Y_ARRAY; y++){
        for(int x = 0; x < SIZE_X_ARRAY; x++){
            printf("%.8f\n", aPtrs[y][x]);
        }
    }
    */

}